src/query/query_hash_scan.c

```diff
@@ -30,10 +40,160 @@
 #include "query_hash_scan.h"
 #include "db_value_printer.hpp"
 #include "dbtype.h"
+#include "chartype.h"
+#include "storage_common.h"
+#include "error_manager.h"
+#include "page_buffer.h"
+#include "slotted_page.h"
+#include "file_manager.h"
+#include "db_date.h"
+#include "thread_compat.hpp"
+#include "oid.h"
+
 
 static bool safe_memcpy (void *data, void *source, int size);
 static DB_VALUE_COMPARE_RESULT qdata_hscan_key_compare (HASH_SCAN_KEY * ckey1, HASH_SCAN_KEY * ckey2, int *diff_pos);
 
+/****************************************************************************/
+/************************ file hash structure *******************************/
+/****************************************************************************/
+/* Number of bits pseudo key consists of */
+#define FHS_HASH_KEY_BITS           (sizeof (FHS_HASH_KEY) * 8)
+
+/* Number of bits the short data type consists of */
+#define FHS_SHORT_BITS  (sizeof (short) * 8)
+
+#define FHS_FLAG_INDIRECT	-1	/* 0xFFFF, record is page id for duplicate key bucket */
+#define FHS_FLAG_DUMMY_NUM	0	/* meaningless flag of record in dk bucket */
+
+#define FHS_HEADER_SLOT_ID	0
+#define FHS_FIRST_SLOT_ID	1
+
+/* key */
+typedef unsigned int FHS_HASH_KEY;
+#define FHS_KEY_TYPE DB_TYPE_INTEGER
+#define FHS_KEY_SIZE (sizeof (FHS_HASH_KEY))
+#define FHS_ALIGNMENT ((char) FHS_KEY_SIZE)
+#define FHS_MAX_DUP_KEY 100	/* 10% of MAXNUM (PAGE 16K / RECORD 14 bytes) */
```

Comment:
I disagree with making it a system parameter. Users will not be able to judge an appropriate that value. Also, since this depends on the data, there will be multiple cases on one system.
The reason I set it to 10% is that it is more important to prevent unnecessary splitting than to waste space in the DK bucket. The performance degradation caused by unnecessary splitting can occur on small data sets.

Reply:
I mean a variable that you want the query optimizer or plan stage to set. I thought it might be possible to adjust it using statistical information in the future.
Also, spliting and moving to the DK page seem to have similar workloads.

However, it does not necessarily need to be corrected, so you can proceed as you wish.

Reply:
thanks. I added comment for later work.